controller/hook.go

```diff
@@ -0,0 +1,148 @@
+	appLabelKey, err := ctrl.settingsMgr.GetAppInstanceLabelKey()
+	if err != nil {
+		return false, err
+	}
+	var revisions []string
+	for _, src := range app.Spec.GetSources() {
+		revisions = append(revisions, src.TargetRevision)
+	}
+
+	targets, _, err := ctrl.appStateManager.GetRepoObjs(app, app.Spec.GetSources(), appLabelKey, revisions, false, false, false, proj)
+	if err != nil {
+		return false, err
+	}
+	runningHooks := map[kube.ResourceKey]*unstructured.Unstructured{}
+	for key, obj := range liveObjs {
+		if isPostDeleteHook(obj) {
+			runningHooks[key] = obj
+		}
+	}
+
+	expectedHook := map[kube.ResourceKey]*unstructured.Unstructured{}
+	for _, obj := range targets {
+		if obj.GetNamespace() == "" {
+			obj.SetNamespace(app.Spec.Destination.Namespace)
+		}
+		if !isPostDeleteHook(obj) {
+			continue
+		}
+		if runningHook := runningHooks[kube.GetResourceKey(obj)]; runningHook == nil {
+			expectedHook[kube.GetResourceKey(obj)] = obj
+		}
+	}
+	createdCnt := 0
+	for _, obj := range expectedHook {
+		_, err = ctrl.kubectl.CreateResource(context.Background(), config, obj.GroupVersionKind(), obj.GetName(), obj.GetNamespace(), obj, v1.CreateOptions{})
+		if err != nil {
+			return false, err
+		}
+		createdCnt++
+	}
+	if createdCnt > 0 {
+		logCtx.Infof("Created %d post-delete hooks", createdCnt)
+		return false, nil
+	}
+	resourceOverrides, err := ctrl.settingsMgr.GetResourceOverrides()
+	if err != nil {
+		return false, err
+	}
+	healthOverrides := lua.ResourceHealthOverrides(resourceOverrides)
+
+	progressingHooksCnt := 0
+	for _, obj := range runningHooks {
+		hookHealth, err := health.GetResourceHealth(obj, healthOverrides)
+		if err != nil {
+			return false, err
+		}
+		if hookHealth.Status == health.HealthStatusProgressing {
+			progressingHooksCnt++
+		}
+	}
+	if progressingHooksCnt > 0 {
+		logCtx.Infof("Waiting for %d post-delete hooks to complete", progressingHooksCnt)
+		return false, nil
+	}
+
+	return true, nil
+}
+
+func (ctrl *ApplicationController) cleanupPostDeleteHooks(liveObjs map[kube.ResourceKey]*unstructured.Unstructured, config *rest.Config, logCtx *log.Entry) (bool, error) {
+	resourceOverrides, err := ctrl.settingsMgr.GetResourceOverrides()
+	if err != nil {
+		return false, err
+	}
+	healthOverrides := lua.ResourceHealthOverrides(resourceOverrides)
+
+	pendingDeletionCount := 0
+	for _, obj := range liveObjs {
+		if !isPostDeleteHook(obj) {
+			continue
+		}
+		hookHealth, err := health.GetResourceHealth(obj, healthOverrides)
+		if err != nil {
+			return false, err
+		}
+		for _, policy := range hook.DeletePolicies(obj) {
+			if policy == common.HookDeletePolicyHookFailed && hookHealth.Status == health.HealthStatusDegraded || policy == common.HookDeletePolicyHookSucceeded && hookHealth.Status == health.HealthStatusHealthy {
```

Comment:
I think this behaviour is different from the delete policy of the other hook types, as far as I remember in gitops-engine the delete policy would be compared with the whole app status instead of just the hook status. We might want to keep the behaviour same or else it can be confusing. I understand at deletion we probably won't have the app status to compare maybe so might be better to have special deletion policy just for the delete hooks?

Reply:
Another question here, what do we do if the deletion policy condition isn't met. Does the app itself remains or is it just the hook that sticks around?

Reply:
You are right. App deletion does not have status, however, I think the use case is similar: the user just needs control to preserve some hook for debugging. To match with gitops engine behavior it makes sense to use aggregated hooks status. Implemented